### PR TITLE
Captain Beepsky and His Merry Medibots

### PR DESCRIPTION
**Pull request by Tastyfish**
Ports tg's simple_animal bot system. There's too many PR's on to cover all changes ported, but here is the "original/main" one https://github.com/tgstation/-tg-station/pull/13239
Bots (securitron, medibot, cleanbot, floorbot, & ED-209) are now simple animal based. What does this mean?
- They can be player-controlled. Currently either by adminbus or by unlocking & opening a bot, and inserting a pAI. There is a simple set of directives bots should follow; if the player is an inserted pAI, they also adopt their pAI directives.
- They can perform all of the tasks of the normal npc bots, and if given orders via PDA are shouted the orders in big text.
- Yes, PC-controlled MULEbots observe the mob avoidance wire.
 - MULEbots actually leave non-shitty tracks now, with DNA, correct color, etc.
- This also fixes all of the MULEbot destinations in the map that were set up wrong (like half of them) including making it physically possible for the MULE to access the engineering one.
- As a tweak, bots can do the silicon emote sounds. `*ping`
- Diagnostic HUD's indicate the health, and operating mode of the bots.
- Bots also get the screen object kind of HUD.
- And more!

MULE

Some miscellaneous simple_animal stuff too that was necessary to make this not a mess...
- Simple animal NPC AI is now handled in a seperate `animal` process, so it can't possibly tie up or crash the main mob one. Just disable it if things go horribly wrong!
- Moved damage and atmos handling to tg's `damage_coeffs` and `atmos_requirements` lists. This makes it much cleaner to define that stuff.
- Also ported over the `healable` and `sentience_type` vars.
- Probably 10 other things I've since forgotten!

Also fixed a NanoUI sanitization bug.

:cl:
rscadd: Bots can now be controlled by players, by opening them up and inserting a pAI. *beep
bugfix: MULEbots can now access the engineering destination.
bugfix: All MULEbot destinations should now have the correct load/unload direction, meaning the crates won't be dumped inside the flaps.
rscadd: Diagnostic HUD's now give health and status information about bots.
bugfix: MULEbot rampage blood tracks are now rendered correctly and handle UE traces as appropriate.
experiment: Bots should (hopefully) lag the game less now.
tweak: Beepsky contains 30% more potato.
/:cl:

### CODE REVIEW

**Comment by KasparoVy** [comment]
pAI controlled ED-209s included?

**Comment by Tastyfish** [comment]
;)

**Comment by NTSAM** [comment]
oh hey now i can have a loyal medibot ass slave that won't fucking move every time someone with the slightest bruise walks in

**Comment by KasparoVy** [comment]
If pAI controlled ED-209s is a thing, all I can really say to that is

"My God JC, a bomb!" :bomb:

**Comment by Spacemanspark** [comment]
Is it possible to get the pAI out of the bot?

**Comment by FalseIncarnate** [comment]
Secbots (beepsky) and ED-209s are not pAI-compatible by default. Only adminbus can change that.

**Comment by Tastyfish** [comment]
@FalseIncarnate I removed the restriction. However, they also have a basic lawset now (that can be overriden at the owner's discretion, but at that point the owner could also have an emag). The restriction can be re-added if it's a problem.

@Spacemanspark Yes. The pAI will also drop if the bot is killed.

**Comment by Spacemanspark** [comment]
Gotcha.
:+1:

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -636,8 +636,8 @@
 // called when something steps onto a human
 // this handles mulebots and vehicles
 /mob/living/carbon/human/Crossed(var/atom/movable/AM)
-	if(istype(AM, /obj/machinery/bot/mulebot))
-		var/obj/machinery/bot/mulebot/MB = AM
+	if(istype(AM, /mob/living/simple_animal/bot/mulebot))
```
Comment (tigercat2000) [inline code/modules/mob/living/carbon/human/human.dm]: why does this use /human/Crossed instead of /mulebot/Move()

Comment (Tastyfish) [inline code/modules/mob/living/carbon/human/human.dm]: Yeah, it could be moved there. Not touching that vehicle code below it though.

**Comment by FalseIncarnate** [comment]
@Tastyfish Ah, it seems to have been changed since when I was looking at the code while writing that comment. Nevermind then, nothing to see here, move along.

**Comment by Aurorablade** [comment]
>  medibot ass slave

....wot...

**Comment by DaveTheHeadcrab** [comment]
> Give medibots an explosive beaker mix
> Insert pAI
> Begin aloha snackbar

**Comment by fludd12** [comment]
Will medipAIs be able to select which chemical they dispense, and do they get a sort of health scanner thing?

Similarly, secbots getting secHUD and the ability to handcuff seperately?

**Comment by Fox-McCloud** [comment]
>....wot...

Better than medibot ass *salve*.

**Comment by Tastyfish** [comment]
@fludd12 Good questions.
Medibots click on a person to heal, and it uses whichever reagent source they're configured for, as if they were the npc.
Medibots get a health HUD and secbots get a security HUD; all bots get a diagnostic HUD.
If a secbot clicks somebody that is not stunned, it stuns them. If they are already stunned, they cuff. So, you could choose to only stun them and stop if you desired.
In case anyone is wondering about the other ones...
ED-209's work similarly to secbots.
Mulebots can drag crates onto themselves, and click an adjacent tile to unload.
Floorbots can click floors to place a tile, and can click floor tiles or raw metal sheets to refill.
All of them get shouted commands in big text if sent PDA orders, but they also get the radio channels appropriate to their department, common, and AI private for orders (I intentionally didn't give them binary for the sake of malfs).

**Comment by FalseIncarnate** [comment]
@Tastyfish Please tell me that secbots and ED-209s can't adjust people's sec status and records via their huds (unless maybe emagged?) so you don't get idiots popping pAIs into bots and having them just go around setting everyone to arrest or messing with sec records just to cause problems as non-traitors.

**Comment by Spacemanspark** [comment]
I don't see any issue with inserting them into any of the securitrons.
...Well, maybe pingsky, but other than that...

**Comment by TheDZD** [comment]
Just curious, but does this fix Beepsky being arresting you if you ever tried to repair him with a welding tool?

**Comment by Tastyfish** [comment]
@TheDZD Yep!

**Comment by Necaladun** [comment]
Woo! As funny as some of the results are, there is really no reason for them to be objs. pAI control seems to have a huge amount of cool features...but...

Idk about Beepsky being player controlled...for some reason it just rubs me as wrong. He doesn't take orders from pAIs. He's a loose cannon who doesn't play by the rules, but gets results!

**Comment by fludd12** [comment]
Maybe have a variable that makes roundstart securitrons not pAI compatible? Beepsky and Pingsky overrides seem a little too painful, for security and malf reasons.

**Comment by Tastyfish** [comment]
Due to feedback, I removed the ability to insert pAI's into the named securitrons.

**Comment by FalseIncarnate** [comment]
:+1: without any issues now. I just don't like the idea of a player taking over Beepsky or Pingsky or Buzzsky, since they are "unique" bots. Especially in the case of Buzzsky, because he's emagged and set to harmbaton at round start, this means an inserted player could ONLY do harm with him (unless that got changed somewhere in the HUGE number of code changes this PR did).

**Comment by theColdflame** [comment]
Can pAIs be inserted into malfunctioning bots?

Player-controled lethal ED would certainly be 'effective'.

**Comment by FlattestGuitar** [comment]
While this PR is happening it would be nice to also get medibots to stop 'healing' IPCs.

Small code change and it kinda fits the theme.

**Comment by ProperPants** [comment]
>While this PR is happening it would be nice to also get medibots to stop 'healing' IPCs.

Small code change and it kinda fits the theme.

While that is happening, could we please change IPC blood types to various industrial fluids and lubricants? Or just WD-40. Or just no blood (unless this would cause issues).

**Comment by Spacemanspark** [comment]
That's a bit out of scope for this PR, @ProperPants

**Comment by Crazylemon64** [comment]
@ProperPants I already made sure species that lacked blood won't bleed anymore, they just have a black tint when "bloodied"

**Comment by Dumbdumn5** [comment]
:heart_eyes_cat:

**Comment by Tastyfish** [comment]
Medibots now completely ignore IPC's.

File: code/modules/mob/living/simple_animal/bot/medbot.dm [hunk 2]
```diff
@@ -353,6 +353,12 @@
 	if(C.suiciding)
 		return 0 //Kevorkian school of robotic medical assistants.
 
+	// is secretly a silicon
+	if(ishuman(C))
+		var/mob/living/carbon/human/H = C
+		if(H.species && H.species.reagent_tag == PROCESS_SYN)
+			return 0
+
```
Comment (Crazylemon64) [inline code/modules/mob/living/simple_animal/bot/medbot.dm]: Ah, good, nice and abstracted, in the manner that makes the most sense.

Comment (Crazylemon64) [inline code/modules/mob/living/simple_animal/bot/medbot.dm]: Actually, you may want to do a check for `& PROCESS_ORG` instead, if we ever add more reagent tags (*scream).

**Comment by Fox-McCloud** [comment]
merge conflicts.

File: code/modules/mob/living/simple_animal/bot/medbot.dm (hunk 2 above)
Comment (Tastyfish) [inline code/modules/mob/living/simple_animal/bot/medbot.dm]: The logic will probably need to be changed for whatever that would mean anyway, heh.

**Comment by Spacemanspark** [comment]
This will be fun.

**Comment by Tastyfish** [comment]
It was conflicts with the disease PR as expected. Resolved

**Comment by Central-Core** [comment]
this looks nice it gives people a reason to be PAI more then just being forced to be held by someone all day.

**Pull request merged**